opentelemetry-api/src/opentelemetry/metrics/meter.py

```diff
@@ -0,0 +1,210 @@
+
+            return wrapper_1
+
+        return wrapper_0
+
+
+class DefaultMeter(Meter):
+    def __init__(self):
+        self._instrument_names = set()
+
+    def _instrument_name_checker(self, name):
+
+        if name in self._instrument_names:
+            raise Exception("Instrument name {} has been used already")
+
+        self._instrument_names.add(name)
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_counter(self, name, unit="", description="") -> Counter:
+        return DefaultCounter(name, unit=unit, description=description)
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_up_down_counter(
+        self, name, unit="", description=""
+    ) -> UpDownCounter:
+        return DefaultUpDownCounter(name, unit=unit, description=description)
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_observable_counter(
+        self, name, callback, unit="", description=""
+    ) -> ObservableCounter:
+        return DefaultObservableCounter(
+            name,
+            callback,
+            unit=unit,
+            description=description,
+        )
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_histogram(self, name, unit="", description="") -> Histogram:
+        return DefaultHistogram(name, unit=unit, description=description)
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_observable_gauge(
+        self, name, callback, unit="", description=""
+    ) -> ObservableGauge:
+        return DefaultObservableGauge(  # pylint: disable=abstract-class-instantiated
+            name,
+            callback,
+            unit=unit,
+            description=description,
+        )
+
+    @Meter.check_unique_name(_instrument_name_checker)
+    def create_observable_up_down_counter(
+        self, name, callback, unit="", description=""
+    ) -> ObservableUpDownCounter:
+        return DefaultObservableUpDownCounter(  # pylint: disable=abstract-class-instantiated
+            name,
+            callback,
+            unit=unit,
+            description=description,
+        )
+
+
+class MeterProvider(ABC):
+    @abstractmethod
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> Meter:
+        pass
+
+
+class DefaultMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> Meter:
+        return DefaultMeter()
+
+
+_METER_PROVIDER = None
+
+
+def set_meter_provider(meter_provider: MeterProvider) -> None:
+    """Sets the current global :class:`~.MeterProvider` object."""
+    global _METER_PROVIDER  # pylint: disable=global-statement
+
+    if _METER_PROVIDER is not None:
+        _logger.warning("Overriding of current MeterProvider is not allowed")
+        return
+
+    _METER_PROVIDER = meter_provider
+
+
+def get_meter_provider() -> MeterProvider:
```

Comment:
I see.  I looked into the current implementation which uses a `ProxyTracerProvider`. This implementation does the job but I wonder if the same can be done without any proxy:

[This](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/tests/trace/test_proxy.py#L39) is the expected flow:

```
# No trace provider has been set, so the user gets a proxy provider
proxy_provider = trace.get_tracer_provider()

# The proxy provider gives the user a proxy tracer
proxy_tracer = provider.get_tracer("proxy-test")

# The spans are started with the proxy tracer
with tracer.start_span("span1") as span:
    self.assertIsInstance(span, trace.NonRecordingSpan)

# The user now sets a real provider
trace.set_tracer_provider(RealProvider())

# References to the old provider return real tracer now
real_tracer = provider.get_tracer("proxy-test")

# Proxy tracers also behave now as real tracers
with tracer.start_span("") as span:
    self.assertIsInstance(span, RealSpan)
```

The point here is that the user is able to call `set_tracer_provider` after `get_tracer_provider` once because before this being called the proxy tracers and tracer providers point to the default ones. We can achieve (almost) the same results if we allow `set_tracer_provider` to be called only once (by simply using a module-level flag, for example).

I say *almost* because with this approach the tracers and tracer providers don't automatically start behaving like something else after `set_tracer_provider` has been called, but the user would have to create new tracer providers and tracers that are now the real ones.

I personally have no problem with using the more magical proxy approach but others may prefer this more obvious approach :shrug:

Reply:
I think the context in #1159 makes it pretty clear we need this. I agree it's kinda magical which I don't like, but sometimes we can't control the import order or the import order may be inconsistent. Maybe we can check with the Java implementation and see how they handle this.

Reply:
We will probably need this for `MeterProvider` as well but can be left for a different PR. @aabmass  can you create an issue?

Reply:
I have used the previous proxy approach: I personally don't have any issues with its "magicality" (is that a word? :thinking: :laughing:), in fact I like it because it provides automatic updating of already created meters and meterproviders. It is also more consistent with what we already have for tracing.